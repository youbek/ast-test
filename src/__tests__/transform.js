/**
 * @jest-environment jsdom
 */

const {
  tokenize,
  buildAST,
  transformFromAST,
  transform,
} = require("../transform");

it("should return correct nested tokens", () => {
  const sourceCode = `
div
    span
    span
div
`;

  const tokens = tokenize(sourceCode);

  expect(tokens).toEqual([
    {
      type: "Tag",
      value: "div",
      indentLevel: 0,
    },
    {
      type: "Tag",
      value: "span",
      indentLevel: 1,
    },
    {
      type: "Tag",
      value: "span",
      indentLevel: 1,
    },
    {
      type: "Tag",
      value: "div",
      indentLevel: 0,
    },
  ]);
});

it("should build correct nested ast tree from given tokens", () => {
  const tokens = [
    {
      type: "Tag",
      value: "div",
      indentLevel: 0,
    },
    {
      type: "Tag",
      value: "span",
      indentLevel: 1,
    },
    {
      type: "Tag",
      value: "span",
      indentLevel: 1,
    },
    {
      type: "Tag",
      value: "div",
      indentLevel: 0,
    },
  ];

  const ast = buildAST(tokens);

  expect(ast).toEqual([
    {
      type: "Element",
      name: "div",
      children: [
        {
          type: "Element",
          name: "span",
          children: [],
        },
        {
          type: "Element",
          name: "span",
          children: [],
        },
      ],
    },
    {
      type: "Element",
      name: "div",
      children: [],
    },
  ]);
});

it("should create correct html from AST tree", () => {
  const ast = [
    {
      type: "Element",
      name: "div",
      children: [
        {
          type: "Element",
          name: "span",
          children: [],
        },
      ],
    },
    {
      type: "Element",
      name: "div",
      children: [],
    },
  ];

  const html = transformFromAST(ast);

  expect(html).toMatchSnapshot();
});

it("should return single element", () => {
  const sourceCode = "div";

  const output = transform(sourceCode);

  expect(output).toBe("<div></div>");
});

it("should return multiple sibling elements", () => {
  const sourceCode = `
div
div
`;

  const output = transform(sourceCode);

  expect(output).toMatchSnapshot();
});

it("should return nested elements", () => {
  const sourceCode = `
div
    span
`;

  const output = transform(sourceCode);

  expect(output).toMatchSnapshot();
});

it("should pass interview question correctly", () => {
  const sourceCode = `
div
    span
    span
    img
div
    div
        div
    div
`;

  const output = transform(sourceCode);
  expect(output).toMatchSnapshot(sourceCode);
});
