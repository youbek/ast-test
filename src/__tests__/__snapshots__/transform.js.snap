// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create correct html from AST tree 1`] = `"<div><span></span></div><div></div>"`;

exports[`should pass interview question correctly: 
div
    span
    span
    img
div
    div
        div
    div
 1`] = `"<div><span></span><span></span><img /></div><div><div><div></div></div><div></div></div>"`;

exports[`should return multiple sibling elements 1`] = `"<div></div><div></div>"`;

exports[`should return nested elements 1`] = `"<div><span></span></div>"`;
